slim_build_context.md

```diff
@@ -0,0 +1,162 @@
+      corner-casey, and like we'll potentially end up discovering and
+      reimplementing various docker logic and features. I only spent an hour
+      digging into this feeling of corner-casiness, and it's possible there's
+      an option that lets us do this robustly. Our [current implementation](https://github.com/windmilleng/tilt/blob/e65742e96b9a7b0684a2e75547cc8089f3ac76ba/internal/dockerfile/dockerfile.go#L147)
+      calls out a few cases. There's also [ARG](https://docs.docker.com/engine/reference/builder/#arg) and
+      [environment replacement](https://docs.docker.com/engine/reference/builder/#environment-replacement)
+      support. We could just implement the 90% cases and worry about corner cases
+      as they're reported, but then they're corner cases of the corner case of
+      this whole spec, which makes them hard to prioritize and leaves the feature
+      feeling buggy regardless.
+
+2. ##### Add a `docker_build` option `only_rebuild_on` - only files listed here will trigger builds, but the build still has access to the full build context.
+
+   Pros:
+
+   1. Simple implementation.
+
+   Cons:
+
+   1. The user has to explicitly list dependencies, which feels like boilerplate/
+      bookkeeping, repeating whatever's in the Dockerfile. Also, as is often the
+      case with such boilerplate, introduces a source of annoyance/user error,
+      in that this might not match the Dockerfile, either because the user fails
+      to keep it in sync with the Dockerfile, or simply typos.
+   2. A simple dependency list isn't very flexible, e.g., what if you want to
+      include go but exclude unit tests, or include js but exclude ts.
+   3. Introduces a distinction between "files we include in the build" and
+      "files we watch to trigger a build", making the user's required
+      mental model a bit more complicated.
+
+3. ##### Add `docker_build` options `ignore` and/or `only`, which filter the build context.
+
+   Pros:
+
+   1. Simple implementation.
+   2. Similar to #2, but more flexible - e.g., can include `internal` and
+      exclude `internal/**/*_test.go`.
+
+   Cons:
+
+   1. Risks feeling boilerplatey, as with option #2 above.
+   2. The flexibility over #2 is nice, but also kind of weird that we'd end up
+      having to do, e.g., `['**', '!package.json']` sometimes (and we don't have
+      good evidence people are very aware of this concept from Dockerfiles,
+      but maybe we can mostly mitigate that by advertising in our docs?). An
+      simpler, less flexible alternative would be `includes=[...], excludes=[...]`.
+      That might be flexible enough.
+
+4. ##### Allow `docker_build`'s context param to take either the string (directory)
+   it currently does, or a new `BuildContext` type that specifies a directory
+   and also watch/build rules.
+
+   Pros:
+
+   1. Simple implementation.
+   2. As a byproduct, might be a good replacement for `deps` in `custom_build`
```

Comment:
[ feel free to ignore for now, but might be a quick win `

`custom_build` is immune from 1 part of the issue this solves for `docker_build`, but suffers from another part:

It's immune from "I only want to watch certain directories", because you can specify certain directories.

It still suffers from "I want to ignore certain files in the directory".

Is it worth adding this to `custom_build`, too? I'd say probably if it's 20% more work ("oh, just need to wire it up from Tiltfile to model, but then the watching code is the same") but not if it's 100% more work ("they're completely separate").